proposals/data-classes.md

```diff
@@ -0,0 +1,241 @@
+
+# Working with Data
+
+When we talk about C# and start to talk about data, the conversation often
+moves to talk about advanced data structures and complicated use cases. Here
+I'd like to do the opposite: talk about simple data and the representations
+we use for it.
+
+To start, let's talk about what data is and what it isn't.
+
+Data *is* a collection of values with potentially heterogenous types. Some
+examples include
+
+* Database rows
+* JSON/XML messages
+* Login info
+* Configuration options
+
+Data *is not*
+
+* A process
+* A computation
+* A conversation
+* Interactive
+* An object
```

Comment:
This _feels_ like a feature better suited to an external code generator rather than an addition to the language.  It is too complicated to be considered syntactic sugar, but doesn't add 'new' functionality, it just "rewraps" existing available features.  A code generator that accepts a data definition and generates the appropriate C# class would provide this without adding complexity to the language.

Reply:
And such generators already exist - here's one: https://github.com/johnazariah/csharp-algebraictypes

Reply:
There are multiple features that could not be implemented by a code generator in this proposal, including the use of object initializers to initialize read-only members.